In [52]:
import pandas as pd
pd.set_option('display.max_rows', 30)

In [39]:
path = '../raw_data/triples.csv'
df = pd.read_csv(path, header=0)

In [53]:
df

head tail               relation label
0     关羽   刘备  younger_sworn_brother    义弟
1     张飞   刘备  younger_sworn_brother    义弟
2     关羽   张飞    elder_sworn_brother    义兄
3     张苞   张飞                    son    儿子
4     关兴   关羽                    son    儿子
5     关平   张苞                  sworn    结拜
6     关平   关羽            adopted_son    义子
7     卢植   刘备                 master    师傅
8    公孙瓒   刘备                 friend    朋友
9     甘氏   刘备                   wife     妻
10    刘禅   甘氏                    son    儿子
11   诸葛瞻   刘禅             son_in_law    女婿
12   诸葛瞻  诸葛亮                    son    儿子
13   诸葛亮   姜维                 master    师傅
14    刘备  诸葛亮                   lord    主公
..   ...  ...                    ...   ...
138  司马昭  司马师        younger_brother    弟弟
139  司马昭  司马懿                    son    儿子
140  司马炎  司马昭                    son    儿子
141   曹仁   曹操              same_clan    同族
142   曹纯   曹仁        younger_brother    弟弟
143   曹昂   曹操                    son    儿子
144   刘氏   曹操                   wife    妻子
145   超昂   刘氏                    son    儿子
146   卞氏   曹操                   wife    妻子
147   曹丕   卞氏                    son    儿子
148   曹植   曹丕        younger_brother    弟弟
149   马谡  诸葛亮              colleague    同事
150   马谡   刘备               minister     臣
151   孙坚   孙权                 father    父亲
152  吴国太   孙权                 mother    母亲

[153 rows x 4 columns]

超昂 应该是曹昂，之后把所有内容都更改一下

145   超昂   刘氏      son    儿子

In [57]:
df.at[145, 'head']

'超昂'

In [58]:
df.at[145, 'head'] = '曹昂'
df.at[145, 'head']

'曹昂'

In [59]:
head_identicle_people = set(df['head'].tolist())
tail_identicle_people = set(df['tail'].tolist())

In [60]:
all_identicle_people = set.union(head_identicle_people, tail_identicle_people)

In [61]:
print(all_identicle_people)
print(len(all_identicle_people))

{'刘禅', '马谡', '貂蝉', '凌统', '公孙越', '蔡瑁', '文聘', '太史慈', '诸葛瞻', '吴国太', '丁原', '大乔', '蒋琬', '张宝', '鲁肃', '王允', '丁奉', '徐庶', '卞氏', '刘表', '夏侯楙', '曹纯', '袁尚', '司马炎', '乐进', '蒋钦', '曹爽', '孟获', '黄盖', '周泰', '马腾', '刘辩', '孙尚香', '黄承彦', '糜氏', '甘氏', '卢植', '郭嘉', '吕布', '黄祖', '刘氏', '袁术', '司马师', '庞统', '曹昂', '法正', '徐晃', '钟会', '曹植', '庞德公', '徐盛', '马良', '张绣', '邓艾', '清河公主', '姜维', '董卓', '曹真', '孙氏', '刘备', '公孙瓒', '诸葛亮', '孙策', '韩当', '曹操', '曹丕', '韩遂', '高顺', '孙韶', '张昭', '黄月英', '曹嵩', '刘胜', '魏延', '张苞', '祝融', '张郃', '马超', '蔡氏', '张辽', '于禁', '诸葛瑾', '周瑜', '关平', '荀彧', '典韦', '袁熙', '关兴', '刘协', '蒯越', '曹仁', '程普', '孙权', '许褚', '夏侯淳', '袁谭', '荀攸', '司马昭', '程昱', '刘启', '庞德', '孙坚', '陈宫', '陆逊', '贾诩', '张飞', '吕蒙', '糜芳', '吕伯奢', '关羽', '糜竺', '张梁', '邹氏', '张角', '甘宁', '黄忠', '袁绍', '沙摩柯', '小乔', '司马懿', '夏侯渊', '赵云'}
122


In [62]:
def find_relation(name):
    p1 = df.loc[df['head'] == name]
    p2 = df.loc[df['tail'] == name]
    print(p1)
    print(p2)

In [63]:
find_relation( '曹昂' )

    head tail relation label
143   曹昂   曹操      son    儿子
145   曹昂   刘氏      son    儿子
Empty DataFrame
Columns: [head, tail, relation, label]
Index: []


In [64]:
find_relation( '超昂' )

Empty DataFrame
Columns: [head, tail, relation, label]
Index: []
Empty DataFrame
Columns: [head, tail, relation, label]
Index: []


夏侯惇被错写为了夏侯淳，更改所有的位置。

In [65]:
find_relation('夏侯淳')

Empty DataFrame
Columns: [head, tail, relation, label]
Index: []
    head tail      relation label
114  夏侯渊  夏侯淳  son_of_uncle    堂弟
115   曹操  夏侯淳     same_clan    同族
116   曹操  夏侯淳          lord    主公


In [66]:
df.at[114, 'tail'] = '夏侯惇'
df.at[115, 'tail'] = '夏侯惇'
df.at[116, 'tail'] = '夏侯惇'

In [67]:
find_relation('夏侯淳')

Empty DataFrame
Columns: [head, tail, relation, label]
Index: []
Empty DataFrame
Columns: [head, tail, relation, label]
Index: []


In [68]:
find_relation('夏侯惇')

Empty DataFrame
Columns: [head, tail, relation, label]
Index: []
    head tail      relation label
114  夏侯渊  夏侯惇  son_of_uncle    堂弟
115   曹操  夏侯惇     same_clan    同族
116   曹操  夏侯惇          lord    主公


一共有122个不同的武将。这里我们将这些武将分为4大势力，蜀国，魏国，吴国，群雄

魏国： '荀攸' '徐晃' '庞德'  '曹操'  '张郃' '荀彧' '司马昭' '曹仁' '曹植' '典韦' '郭嘉' '乐进' '张辽' '夏侯淳' '曹真' '曹爽' '曹纯' '司马炎' '司马懿' '邓艾' '夏侯渊' '贾诩' '许褚' '司马师' '夏侯楙' '蔡瑁' '曹嵩' '邹氏' '程昱' '刘氏' '清河公主' '于禁' '蒯越' '钟会' '文聘' '曹昂'
 
蜀国：'关兴' '魏延' '诸葛亮' '诸葛瞻' '法正' '黄忠' '关羽' '赵云' '姜维' '张飞' '徐庶' '马良' '诸葛瑾'  '糜芳'  '刘禅' '黄月英' '庞统' '马超' '祝融'  '黄承彦' '沙摩柯' '甘氏' '糜竺' '刘备' '关平' '张苞' '糜氏' '蒋琬' '马谡' 
 
吴国：'孙坚' '太史慈' '孙策' '大乔' '周泰' '孙权' '黄盖' '周瑜' '甘宁' '吴国太'  '吕蒙' '曹丕' '曹昂' '鲁肃'  '孙尚香' '陆逊' '程普' '徐盛' '孙氏'  '韩当' '张昭' '孙韶' '蒋钦' '凌统' '丁奉' '小乔'
 
 
群雄：'袁术' '马腾' '陈宫' '公孙瓒'  '吕布'  '袁绍' '董卓' '孟获' '貂蝉' '吕伯奢' '刘辩' '韩遂' '丁原' '张角' '袁谭' '刘胜' '公孙越' '张梁' '张宝'  '庞德公' '袁熙'  '刘表'  '卢植' '刘协' '王允' '袁尚' '高顺' '张绣' '刘启' '卞氏' '黄祖' '蔡氏'

In [69]:
def string2list(country):
    """convert name string to name list"""
    country = country.replace("'", '')
    return country.split()

In [70]:
wei = "'荀攸' '徐晃' '庞德'  '曹操'  '张郃' '荀彧' '司马昭' '曹仁' '曹植' '典韦' '郭嘉' '乐进' '张辽' '夏侯淳' '曹真' '曹爽' '曹纯' '司马炎' '司马懿' '邓艾' '夏侯渊' '贾诩' '许褚' '司马师' '夏侯楙' '蔡瑁' '曹嵩' '邹氏' '程昱' '刘氏' '清河公主' '于禁' '蒯越' '钟会' '文聘' '曹昂'"
shu = "'关兴' '魏延' '诸葛亮' '诸葛瞻' '法正' '黄忠' '关羽' '赵云' '姜维' '张飞' '徐庶' '马良' '诸葛瑾'  '糜芳'  '刘禅' '黄月英' '庞统' '马超' '祝融'  '黄承彦' '沙摩柯' '甘氏' '糜竺' '刘备' '关平' '张苞' '糜氏' '蒋琬' '马谡'"
wu = "'孙坚' '太史慈' '孙策' '大乔' '周泰' '孙权' '黄盖' '周瑜' '甘宁' '吴国太'  '吕蒙' '曹丕' '曹昂' '鲁肃'  '孙尚香' '陆逊' '程普' '徐盛' '孙氏'  '韩当' '张昭' '孙韶' '蒋钦' '凌统' '丁奉' '小乔'"
others = "'袁术' '马腾' '陈宫' '公孙瓒'  '吕布'  '袁绍' '董卓' '孟获' '貂蝉' '吕伯奢' '刘辩' '韩遂' '丁原' '张角' '袁谭' '刘胜' '公孙越' '张梁' '张宝'  '庞德公' '袁熙'  '刘表'  '卢植' '刘协' '王允' '袁尚' '高顺' '张绣' '刘启' '卞氏' '黄祖' '蔡氏'"

In [71]:
wei = string2list(wei)
shu = string2list(shu)
wu = string2list(wu)
others = string2list(others)

In [72]:
len(wei) + len(shu) + len(wu) + len(others)

123

到此为止得到了保存有三个势力的

In [73]:
four_camp = wei + shu + wu + others 
len(four_camp)

123

four_camp的总人数是123人，说明有重复的一人。下面写个函数，来找到是哪个人加入了两个势力。

In [74]:
for people in four_camp:
    count = four_camp.count(people)
    if count == 2:
        print(people)

曹昂
曹昂


好的，发现是曹昂加入了吴国阵营。。下面重新设定一下人物列表。为了方便给之后的pandas添加新的一列，这里用dict来保存所有人的配属：

In [75]:
wei = "'荀攸' '徐晃' '庞德'  '曹操'  '张郃' '荀彧' '司马昭' '曹仁' '曹植' '典韦' '郭嘉' '乐进' '张辽' '夏侯惇' '曹真' '曹爽' '曹纯' '司马炎' '司马懿' '邓艾' '夏侯渊' '贾诩' '许褚' '司马师' '夏侯楙' '蔡瑁' '曹嵩' '邹氏' '程昱' '刘氏' '清河公主' '于禁' '蒯越' '钟会' '文聘' '曹昂'"
shu = "'关兴' '魏延' '诸葛亮' '诸葛瞻' '法正' '黄忠' '关羽' '赵云' '姜维' '张飞' '徐庶' '马良' '诸葛瑾'  '糜芳'  '刘禅' '黄月英' '庞统' '马超' '祝融'  '黄承彦' '沙摩柯' '甘氏' '糜竺' '刘备' '关平' '张苞' '糜氏' '蒋琬' '马谡'"
wu = "'孙坚' '太史慈' '孙策' '大乔' '周泰' '孙权' '黄盖' '周瑜' '甘宁' '吴国太'  '吕蒙' '曹丕' '鲁肃'  '孙尚香' '陆逊' '程普' '徐盛' '孙氏'  '韩当' '张昭' '孙韶' '蒋钦' '凌统' '丁奉' '小乔'"
others = "'袁术' '马腾' '陈宫' '公孙瓒'  '吕布'  '袁绍' '董卓' '孟获' '貂蝉' '吕伯奢' '刘辩' '韩遂' '丁原' '张角' '袁谭' '刘胜' '公孙越' '张梁' '张宝'  '庞德公' '袁熙'  '刘表'  '卢植' '刘协' '王允' '袁尚' '高顺' '张绣' '刘启' '卞氏' '黄祖' '蔡氏'"

In [76]:
def assign_group(people_group, group, group_name=''):
    group = group.replace("'", '')
    group = group.split()
    for people in group:
        people_group[people] = group_name
    return people_group

In [77]:
people_group = {}
people_group = assign_group(people_group, wei, group_name='魏国')
people_group = assign_group(people_group, shu, group_name='蜀国')
people_group = assign_group(people_group, wu, group_name='吴国')
people_group = assign_group(people_group, others, group_name='群雄')

In [78]:
people_group

{'丁原': '群雄',
 '丁奉': '吴国',
 '乐进': '魏国',
 '于禁': '魏国',
 '公孙瓒': '群雄',
 '公孙越': '群雄',
 '关兴': '蜀国',
 '关平': '蜀国',
 '关羽': '蜀国',
 '典韦': '魏国',
 '凌统': '吴国',
 '刘协': '群雄',
 '刘启': '群雄',
 '刘备': '蜀国',
 '刘氏': '魏国',
 '刘禅': '蜀国',
 '刘胜': '群雄',
 '刘表': '群雄',
 '刘辩': '群雄',
 '卞氏': '群雄',
 '卢植': '群雄',
 '司马师': '魏国',
 '司马懿': '魏国',
 '司马昭': '魏国',
 '司马炎': '魏国',
 '吕伯奢': '群雄',
 '吕布': '群雄',
 '吕蒙': '吴国',
 '吴国太': '吴国',
 '周泰': '吴国',
 '周瑜': '吴国',
 '夏侯惇': '魏国',
 '夏侯楙': '魏国',
 '夏侯渊': '魏国',
 '大乔': '吴国',
 '太史慈': '吴国',
 '姜维': '蜀国',
 '孙坚': '吴国',
 '孙尚香': '吴国',
 '孙权': '吴国',
 '孙氏': '吴国',
 '孙策': '吴国',
 '孙韶': '吴国',
 '孟获': '群雄',
 '小乔': '吴国',
 '庞德': '魏国',
 '庞德公': '群雄',
 '庞统': '蜀国',
 '张宝': '群雄',
 '张昭': '吴国',
 '张梁': '群雄',
 '张绣': '群雄',
 '张苞': '蜀国',
 '张角': '群雄',
 '张辽': '魏国',
 '张郃': '魏国',
 '张飞': '蜀国',
 '徐庶': '蜀国',
 '徐晃': '魏国',
 '徐盛': '吴国',
 '文聘': '魏国',
 '曹丕': '吴国',
 '曹仁': '魏国',
 '曹嵩': '魏国',
 '曹操': '魏国',
 '曹昂': '魏国',
 '曹植': '魏国',
 '曹爽': '魏国',
 '曹真': '魏国',
 '曹纯': '魏国',
 '沙摩柯': '蜀国',
 '法正': '蜀国',
 '清河公主': '魏国',
 '王允': '群雄',
 '甘宁': '吴国',
 '甘氏': '

In [79]:
len(people_group)

122

人物数目一致。接下来创建两列，给每行的两个任务添加所属的阵营。直接使用map

In [80]:
df['head_group'] = df['head'].map(lambda x:people_group[x])
df['tail_group'] = df['tail'].map(lambda x:people_group[x])

In [81]:
df

head tail               relation label head_group tail_group
0     关羽   刘备  younger_sworn_brother    义弟         蜀国         蜀国
1     张飞   刘备  younger_sworn_brother    义弟         蜀国         蜀国
2     关羽   张飞    elder_sworn_brother    义兄         蜀国         蜀国
3     张苞   张飞                    son    儿子         蜀国         蜀国
4     关兴   关羽                    son    儿子         蜀国         蜀国
5     关平   张苞                  sworn    结拜         蜀国         蜀国
6     关平   关羽            adopted_son    义子         蜀国         蜀国
7     卢植   刘备                 master    师傅         群雄         蜀国
8    公孙瓒   刘备                 friend    朋友         群雄         蜀国
9     甘氏   刘备                   wife     妻         蜀国         蜀国
10    刘禅   甘氏                    son    儿子         蜀国         蜀国
11   诸葛瞻   刘禅             son_in_law    女婿         蜀国         蜀国
12   诸葛瞻  诸葛亮                    son    儿子         蜀国         蜀国
13   诸葛亮   姜维                 master    师傅         蜀国         蜀国
14    刘备  诸葛亮                   lord    主公         蜀国         蜀国
..   ...  ...                    ...   ...        ...        ...
138  司马昭  司马师        younger_brother    弟弟         魏国         魏国
139  司马昭  司马懿                    son    儿子         魏国         魏国
140  司马炎  司马昭                    son    儿子         魏国         魏国
141   曹仁   曹操              same_clan    同族         魏国         魏国
142   曹纯   曹仁        younger_brother    弟弟         魏国         魏国
143   曹昂   曹操                    son    儿子         魏国         魏国
144   刘氏   曹操                   wife    妻子         魏国         魏国
145   曹昂   刘氏                    son    儿子         魏国         魏国
146   卞氏   曹操                   wife    妻子         群雄         魏国
147   曹丕   卞氏                    son    儿子         吴国         群雄
148   曹植   曹丕        younger_brother    弟弟         魏国         吴国
149   马谡  诸葛亮              colleague    同事         蜀国         蜀国
150   马谡   刘备               minister     臣         蜀国         蜀国
151   孙坚   孙权                 father    父亲         吴国         吴国
152  吴国太   孙权                 mother    母亲         吴国         吴国

[153 rows x 6 columns]

最后把relation这一列去掉就行了

In [82]:
df = df.drop(['relation'], axis=1)

In [83]:
df

head tail label head_group tail_group
0     关羽   刘备    义弟         蜀国         蜀国
1     张飞   刘备    义弟         蜀国         蜀国
2     关羽   张飞    义兄         蜀国         蜀国
3     张苞   张飞    儿子         蜀国         蜀国
4     关兴   关羽    儿子         蜀国         蜀国
5     关平   张苞    结拜         蜀国         蜀国
6     关平   关羽    义子         蜀国         蜀国
7     卢植   刘备    师傅         群雄         蜀国
8    公孙瓒   刘备    朋友         群雄         蜀国
9     甘氏   刘备     妻         蜀国         蜀国
10    刘禅   甘氏    儿子         蜀国         蜀国
11   诸葛瞻   刘禅    女婿         蜀国         蜀国
12   诸葛瞻  诸葛亮    儿子         蜀国         蜀国
13   诸葛亮   姜维    师傅         蜀国         蜀国
14    刘备  诸葛亮    主公         蜀国         蜀国
..   ...  ...   ...        ...        ...
138  司马昭  司马师    弟弟         魏国         魏国
139  司马昭  司马懿    儿子         魏国         魏国
140  司马炎  司马昭    儿子         魏国         魏国
141   曹仁   曹操    同族         魏国         魏国
142   曹纯   曹仁    弟弟         魏国         魏国
143   曹昂   曹操    儿子         魏国         魏国
144   刘氏   曹操    妻子         魏国         魏国
145   曹昂   刘氏    儿子         魏国         魏国
146   卞氏   曹操    妻子         群雄         魏国
147   曹丕   卞氏    儿子         吴国         群雄
148   曹植   曹丕    弟弟         魏国         吴国
149   马谡  诸葛亮    同事         蜀国         蜀国
150   马谡   刘备     臣         蜀国         蜀国
151   孙坚   孙权    父亲         吴国         吴国
152  吴国太   孙权    母亲         吴国         吴国

[153 rows x 5 columns]

输出所有的关系，之后针对这些关系设置一些近义词方便检索

In [30]:
set(df.label.values)

{'主公',
 '义兄',
 '义女',
 '义子',
 '义弟',
 '义父',
 '儿子',
 '叔叔',
 '同事',
 '同族',
 '哥哥',
 '堂弟',
 '夫',
 '女儿',
 '女婿',
 '妹妹',
 '妻',
 '妻子',
 '妾',
 '嫂子',
 '岳母',
 '师傅',
 '弟弟',
 '徒弟',
 '朋友',
 '母亲',
 '父亲',
 '结拜',
 '臣',
 '让位',
 '诸侯'}

# 很不幸发现有几行是重复的

In [84]:
def find_relation(name):
    p1 = df.loc[df['head'] == name]
    p2 = df.loc[df['tail'] == name]
    print(p1)
    print(p2)

In [85]:
find_relation( '刘禅' )

   head tail label head_group tail_group
10   刘禅   甘氏    儿子         蜀国         蜀国
29   刘禅   甘氏    儿子         蜀国         蜀国
   head tail label head_group tail_group
11  诸葛瞻   刘禅    女婿         蜀国         蜀国
30  诸葛瞻   刘禅    女婿         蜀国         蜀国


In [86]:
df

head tail label head_group tail_group
0     关羽   刘备    义弟         蜀国         蜀国
1     张飞   刘备    义弟         蜀国         蜀国
2     关羽   张飞    义兄         蜀国         蜀国
3     张苞   张飞    儿子         蜀国         蜀国
4     关兴   关羽    儿子         蜀国         蜀国
5     关平   张苞    结拜         蜀国         蜀国
6     关平   关羽    义子         蜀国         蜀国
7     卢植   刘备    师傅         群雄         蜀国
8    公孙瓒   刘备    朋友         群雄         蜀国
9     甘氏   刘备     妻         蜀国         蜀国
10    刘禅   甘氏    儿子         蜀国         蜀国
11   诸葛瞻   刘禅    女婿         蜀国         蜀国
12   诸葛瞻  诸葛亮    儿子         蜀国         蜀国
13   诸葛亮   姜维    师傅         蜀国         蜀国
14    刘备  诸葛亮    主公         蜀国         蜀国
..   ...  ...   ...        ...        ...
138  司马昭  司马师    弟弟         魏国         魏国
139  司马昭  司马懿    儿子         魏国         魏国
140  司马炎  司马昭    儿子         魏国         魏国
141   曹仁   曹操    同族         魏国         魏国
142   曹纯   曹仁    弟弟         魏国         魏国
143   曹昂   曹操    儿子         魏国         魏国
144   刘氏   曹操    妻子         魏国         魏国
145   曹昂   刘氏    儿子         魏国         魏国
146   卞氏   曹操    妻子         群雄         魏国
147   曹丕   卞氏    儿子         吴国         群雄
148   曹植   曹丕    弟弟         魏国         吴国
149   马谡  诸葛亮    同事         蜀国         蜀国
150   马谡   刘备     臣         蜀国         蜀国
151   孙坚   孙权    父亲         吴国         吴国
152  吴国太   孙权    母亲         吴国         吴国

[153 rows x 5 columns]

In [87]:
# 删除重复行
df = df.drop_duplicates()

In [89]:
df

head tail label head_group tail_group
0     关羽   刘备    义弟         蜀国         蜀国
1     张飞   刘备    义弟         蜀国         蜀国
2     关羽   张飞    义兄         蜀国         蜀国
3     张苞   张飞    儿子         蜀国         蜀国
4     关兴   关羽    儿子         蜀国         蜀国
5     关平   张苞    结拜         蜀国         蜀国
6     关平   关羽    义子         蜀国         蜀国
7     卢植   刘备    师傅         群雄         蜀国
8    公孙瓒   刘备    朋友         群雄         蜀国
9     甘氏   刘备     妻         蜀国         蜀国
10    刘禅   甘氏    儿子         蜀国         蜀国
11   诸葛瞻   刘禅    女婿         蜀国         蜀国
12   诸葛瞻  诸葛亮    儿子         蜀国         蜀国
13   诸葛亮   姜维    师傅         蜀国         蜀国
14    刘备  诸葛亮    主公         蜀国         蜀国
..   ...  ...   ...        ...        ...
138  司马昭  司马师    弟弟         魏国         魏国
139  司马昭  司马懿    儿子         魏国         魏国
140  司马炎  司马昭    儿子         魏国         魏国
141   曹仁   曹操    同族         魏国         魏国
142   曹纯   曹仁    弟弟         魏国         魏国
143   曹昂   曹操    儿子         魏国         魏国
144   刘氏   曹操    妻子         魏国         魏国
145   曹昂   刘氏    儿子         魏国         魏国
146   卞氏   曹操    妻子         群雄         魏国
147   曹丕   卞氏    儿子         吴国         群雄
148   曹植   曹丕    弟弟         魏国         吴国
149   马谡  诸葛亮    同事         蜀国         蜀国
150   马谡   刘备     臣         蜀国         蜀国
151   孙坚   孙权    父亲         吴国         吴国
152  吴国太   孙权    母亲         吴国         吴国

[146 rows x 5 columns]

In [88]:
path = '../raw_data/triples_processed.txt'
df.to_csv(path, header=None, index=None, sep=',')